subprojects/language-native/src/integTest/groovy/org/gradle/language/nativeplatform/ParallelNativePluginsIntegrationTest.groovy

```diff
@@ -140,5 +146,149 @@ class ParallelNativePluginsIntegrationTest extends AbstractInstalledToolChainInt
+        succeeds("assemble")
+
+        then:
+        assertTasksAreParallel("create.*StaticLibrary")
+    }
+
+    def "can execute compile tasks in parallel"() {
+        given:
+        withComponentsForApps(apps)
+
+        when:
+        succeeds("assemble")
+
+        then:
+        assertTasksAreParallel("compile.*")
+    }
+
+    def "can execute assemble tasks in parallel"() {
+        given:
+        apps = [
+            assemblerWithC1 : new AssemblerWithCHelloWorldApp(toolChain),
+            assemblerWithC2 : new AssemblerWithCHelloWorldApp(toolChain)
+        ]
+        withComponentsForApps(apps)
+
+        when:
+        succeeds("assemble")
+
+        then:
+        assertTasksAreParallel("assembleAssembler.*")
+    }
+
+    @RequiresInstalledToolChain(VISUALCPP)
+    def "can execute windows resource compile tasks in parallel"() {
+        given:
+        apps = [
+            windowsRC1 : new WindowsResourceHelloWorldApp(),
+            windowsRC2 : new WindowsResourceHelloWorldApp()
+        ]
+        withComponentsForApps(apps)
+
+        when:
+        succeeds("assemble")
+
+        then:
+        assertTasksAreParallel("compile.*Rc")
+    }
+
+    /**
+     * This is an approximation to check that a set of tasks execute in parallel by default.  Basically it checks to
+     * see if any of the tasks execute concurrently with any other task.  We can't guarantee that another task will
+     * always run concurrently while a given task is executing, so we settle for checking that the opposite does not
+     * happen (i.e. that none of the tasks execute concurrently with other tasks).
+     */
+    void assertTasksAreParallel(String regex) {
+        def tasks = buildOperations.all(ExecuteTaskBuildOperationType, new Spec<BuildOperationRecord>() {
+            @Override
+            boolean isSatisfiedBy(BuildOperationRecord record) {
+                return record.displayName.matches("Task :${regex}")
+            }
+        })
+        assert tasks.size() == apps.size()
+        assert tasks.any { buildOperations.getOperationsConcurrentWith(ExecuteTaskBuildOperationType, it).size() > 0 }
+    }
+
+    def withComponentsForApps(Map<String, TestApp> apps) {
+        apps.each { name, app ->
+            buildFile << app.pluginScript
+            buildFile << app.getExtraConfiguration("${name}Executable")
+            buildFile << """
+                model {
+                    components {
+                        ${name}(NativeExecutableSpec)
+                    }
+                }
+            """
+
+            app.writeSources(file("src/$name"))
+        }
+    }
+
+    def withComponentsForAppsAndStaticLibs(Map<String, TestApp> apps) {
+        withComponentsForAppsAndLibs(apps, true)
+    }
+
+    def withComponentsForAppsAndSharedLibs(Map<String, TestApp> apps) {
+        withComponentsForAppsAndLibs(apps, false)
+    }
+
+    def withComponentsForAppsAndLibs(Map<String, TestApp> apps, boolean useStaticLibs) {
```

Comment:
Would accepting a String here (`static` or `shared`) improve the readability of the function above?